### ISSUE TITLE
Getting "Cannot read properties of undefined" errors when reading from Relay store

### ISSUE THREAD

**Issue by arcticmatt** [facebook/relay#3932]
In Sentry, I'm seeing a lot of issues like this

<img width="1955" alt="image" src="https://user-images.githubusercontent.com/4356178/171468429-137ca7cb-a035-4d93-878d-c747163b702c.png">

The corresponding code looks like this:

```
return (
  <div>
    {data.metadataAccountsForExplore.metadataAccounts.edges.map(
      ({ node: metadataAccount }) => (
        ...
      )
    )}
  </div>
);
```

This is unexpected, because the Relay compiler generates a non-nullable type for `data.metadataAccountsForExplore`. 

Most of these errors involve a user navigating from `/explore` to another page, and then back to `/explore`. As you can see in the Sentry screenshot (first screenshot), there are no additional queries kicked off when the user navigates back to `/explore`. This suggests the issue may be related to reading from the Relay store.

This hypothesis is at odds with [Relay's documentation](https://relay.dev/docs/guided-tour/reusing-cached-data/fetch-policies/) though:

> By default, Relay will try to read the query from the local cache; if any piece of data for that query is [missing](https://relay.dev/docs/guided-tour/reusing-cached-data/presence-of-data/) or [stale](https://relay.dev/docs/guided-tour/reusing-cached-data/staleness-of-data/), it will fetch the entire query from the network. This default fetchPolicy is called "store-or-network".

So basically, I have no idea why this is happening.

**Comment by captbaritone** [comment]
I have an internal blog post that outlines some reasons that fields can be unexpectedly null. I've redacted the internal bits and posted it here: https://gist.github.com/captbaritone/eb7e95d64ac6de71e8ab5d4483e6acf1

Let me know if any of those edge cases end up being the cause of your issue.

**Comment by arcticmatt** [comment]
@captbaritone I don't think it's any of those? Here's a bit more info on what's happening

# With `gcReleaseBufferSize = 10`

The same query gets called with different inputs when I switch between tabs in our UI. Here, you can see all the data is correctly typed as `MetadataAccountsConnection`

<img width="1610" alt="image" src="https://user-images.githubusercontent.com/4356178/172014844-8a695333-4af2-40e9-bc41-1117223f0812.png">

# With `gcReleaseBufferSize = 1`

At first, the store looks fine

<img width="1616" alt="image" src="https://user-images.githubusercontent.com/4356178/172015142-a81f7128-9756-411c-af65-0895b9e5036a.png">

But as more queries get sent, some things get typed as `Object`

<img width="1611" alt="image" src="https://user-images.githubusercontent.com/4356178/172015167-093ffe9f-051b-484a-ae43-48c860cd47b8.png">

Unlike the entries typed as `MetadataAccountsConnection`, I cannot click into the Object's `__ref`

<img width="1608" alt="image" src="https://user-images.githubusercontent.com/4356178/172015194-c19801f3-54b4-4508-9c75-10c35bffa08e.png">

Eventually, more and more stuff gets typed as `Object`, and that's when I run into the failure

<img width="1614" alt="image" src="https://user-images.githubusercontent.com/4356178/172015285-e6410e3a-3d6c-4c9a-a295-9135e35e555a.png">

**Comment by vxm5091** [comment]
@arcticmatt I'm running into a very similar issue - did you end up figuring out the root cause?

**Comment by alexus37** [comment]
We are still encountering this issue. @captbaritone, is there a recommended way to prevent this?

**Comment by alexus37** [comment]
In the event that someone arrives at this location, we have determined that the issue in our situation was the reuse of entry points containing disposed query references.
